flask_admin/model/template.py

```diff
@@ -1,3 +1,115 @@
+from jinja2 import contextfunction
+
+from flask_admin._compat import string_types, reduce
+from flask_admin.babel import gettext
+
+
+class BaseListRowAction(object):
+    def __init__(self, title=None):
+        self.title = title
+
+    def render(self, context, row_id, row):
+        raise NotImplementedError()
+
+    @contextfunction
+    def render_ctx(self, context, row_id, row):
+        return self.render(context, row_id, row)
+
+    def _resolve_symbol(self, context, symbol):
+        if '.' in symbol:
+            parts = symbol.split('.')
+            m = context.resolve(parts[0])
+            return reduce(getattr, parts[1:], m)
+        else:
+            return context.resolve(symbol)
+
+
+class LinkRowAction(BaseListRowAction):
+    def __init__(self, icon_class, url, title=None):
+        super(LinkRowAction, self).__init__(title=title)
+
+        self.url = url
+        self.icon_class = icon_class
+
+    def render(self, context, row_id, row):
+        m = self._resolve_symbol(context, 'row_actions.link')
+
+        if isinstance(self.url, string_types):
+            url = self.url.format(row_id=row_id)
+        else:
+            url = self.url(self, row_id, row)
+
+        return m(self, url)
+
+
+class EndpointLinkRowAction(BaseListRowAction):
+    def __init__(self, icon_class, endpoint, title=None, id_arg='id', url_args=None):
+        super(EndpointLinkRowAction, self).__init__(title=title)
+
+        self.icon_class = icon_class
+        self.endpoint = endpoint
+        self.id_arg = id_arg
+        self.url_args = url_args
+
+    def render(self, context, row_id, row):
+        m = self._resolve_symbol(context, 'row_actions.link')
+        get_url = self._resolve_symbol(context, 'get_url')
+
+        kwargs = dict(self.url_args) if self.url_args else {}
+        kwargs[self.id_arg] = row_id
+
+        url = get_url(self.endpoint, **kwargs)
+
+        return m(self, url)
+
+
+class TemplateLinkRowAction(BaseListRowAction):
+    def __init__(self, template_name, title=None):
+        super(TemplateLinkRowAction, self).__init__(title=title)
+
+        self.template_name = template_name
+
+    def render(self, context, row_id, row):
+        m = self._resolve_symbol(context, self.template_name)
+        return m(self, row_id, row)
+
+
+class ViewRowAction(TemplateLinkRowAction):
+    def __init__(self):
+        super(ViewRowAction, self).__init__(
+            'row_actions.view_row',
+            gettext('View row'))
```

Comment:
I ended up changing this to "View Record" to match with the text we already had translated: https://github.com/flask-admin/flask-admin/commit/882cdf0eeb0f15564d6735b31e57d2ce8704c00d

Also changed "Edit row" to "Edit Record".